src/mon/MgrMonitor.cc

```diff
@@ -70,63 +70,33 @@ static ostream& _prefix(std::ostream *_dout, Monitor &mon,
 
 // the system treats always_on_modules as if they provide built-in functionality
 // by ensuring that they are always enabled.
-const static std::map<uint32_t, std::set<std::string>> always_on_modules = {
-  {
-    CEPH_RELEASE_OCTOPUS, {
-      "crash",
-      "status",
-      "progress",
-      "balancer",
-      "devicehealth",
-      "orchestrator",
-      "rbd_support",
-      "volumes",
-      "pg_autoscaler",
-      "telemetry",
-    }
-  },
-  {
-    CEPH_RELEASE_PACIFIC, {
-      "crash",
-      "status",
-      "progress",
-      "balancer",
-      "devicehealth",
-      "orchestrator",
-      "rbd_support",
-      "volumes",
-      "pg_autoscaler",
-      "telemetry",
-    }
-  },
-  {
-    CEPH_RELEASE_QUINCY, {
-      "crash",
-      "status",
-      "progress",
-      "balancer",
-      "devicehealth",
-      "orchestrator",
-      "rbd_support",
-      "volumes",
-      "pg_autoscaler",
-      "telemetry",
-    }
-  },
-  {
-    CEPH_RELEASE_REEF, {
+static const std::map<uint32_t, std::set<std::string>>& release_always_on_modules() {
+  static std::map<uint32_t, std::set<std::string>> always_on_modules;
+  static std::once_flag did_init;
+  std::call_once(did_init, []() {
+    std::set<std::string> octopus_modules = {
       "crash",
       "status",
       "progress",
       "balancer",
       "devicehealth",
       "orchestrator",
-      "rbd_support",
-      "volumes",
       "pg_autoscaler",
       "telemetry",
-    }
-  },
+    };
+#   ifdef WITH_CEPHFS
+      octopus_modules.insert("volumes");
+#   endif
+#   ifdef WITH_RBD
+      octopus_modules.insert("rbd_support");
+#   endif
+
+    always_on_modules[CEPH_RELEASE_OCTOPUS] = octopus_modules;
+    always_on_modules[CEPH_RELEASE_PACIFIC] = octopus_modules;
+    always_on_modules[CEPH_RELEASE_QUINCY] = octopus_modules;
+    always_on_modules[CEPH_RELEASE_REEF] = octopus_modules;
```

Comment:
Nit: naming a common set of modules after a particular release is a bit confusing -- one immediately wonders whether ` always_on_modules[CEPH_RELEASE_PACIFIC] = octopus_modules` is a cut-and-paste typo.  I would suggest `common_modules` or `always_on_modules` for the set and `always_on_modules_map` for the "release" map.

Reply:
I thought about this a little and chose this convention to emphasize the fact that the set is the same for all releases. Now that you mention it I agree that it may appear confusing more than telling.

TBH, I'm a little disappointed by the redundancy of this structure. I'd only introduce a new set when it is really different, and look up by the `lower_bound` of the current release.